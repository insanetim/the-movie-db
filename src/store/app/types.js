export const SHOW_MODAL = 'app/SHOW_MODAL'
export const HIDE_MODAL = 'app/HIDE_MODAL'
export const SHOW_NOTIFICATION = 'app/SHOW_NOTIFICATION'
export const HIDE_NOTIFICATION = 'app/HIDE_NOTIFICATION'
