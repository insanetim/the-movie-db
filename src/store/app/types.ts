import type { ModalProps, ModalTypes } from 'src/components/ModalRoot/types'
import type { NOTIFICATION_TYPE } from 'src/constants/app'

export type IModal = {
  modalProps: ModalProps | null
  modalType: ModalTypes | null
}

export type INotification = {
  duration: number
  id: string
  messageText: string
  messageType: NOTIFICATION_TYPE
}

export interface AppState {
  modal: IModal
  notifications: INotification[]
}
