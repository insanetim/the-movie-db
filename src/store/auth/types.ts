import { IAccount } from 'src/interfaces/account.interface'

export type UserData = {
  password: string
  username: string
}

export type RequestToken = {
  expires_at: string
  request_token: string
  success: boolean
}

export type Session = {
  session_id: string
  success: boolean
}

export type AuthState = {
  account: IAccount | null
  isAuthenticated: boolean
}
