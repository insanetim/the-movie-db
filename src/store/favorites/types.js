export const FETCH_FAVORITES = 'watchlist/FETCH_FAVORITES'
export const FETCH_FAVORITES_REQUEST = 'watchlist/FETCH_FAVORITES_REQUEST'
export const FETCH_FAVORITES_SUCCESS = 'watchlist/FETCH_FAVORITES_SUCCESS'
export const FETCH_FAVORITES_FAILURE = 'watchlist/FETCH_FAVORITES_FAILURE'
export const SET_FAVORITES_PAGE = 'watchlist/SET_FAVORITES_PAGE'
