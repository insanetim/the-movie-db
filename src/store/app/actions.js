import { v4 as uuid } from 'uuid'

import { NOTIFICATION_DURATION, NOTIFICATION_TYPE } from 'src/constants'

import * as types from './types'

export const showModal = ({ modalType, modalProps }) => ({
  type: types.SHOW_MODAL,
  payload: { modalType, modalProps }
})

export const hideModal = () => ({
  type: types.HIDE_MODAL
})

export const showNotification = ({
  id = uuid(),
  messageType = NOTIFICATION_TYPE.SUCCESS,
  messageText,
  duration = NOTIFICATION_DURATION
}) => ({
  type: types.SHOW_NOTIFICATION,
  payload: {
    id,
    messageType,
    messageText,
    duration
  }
})

export const hideNotification = id => ({
  type: types.HIDE_NOTIFICATION,
  payload: id
})
