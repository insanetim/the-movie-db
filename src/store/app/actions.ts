import { createAction, nanoid } from '@reduxjs/toolkit'
import { NOTIFICATION_DURATION, NOTIFICATION_TYPE } from 'src/constants/app'

import type { INotification, ShowModalProps, ShowNotificationProps } from './types'

import { HIDE_MODAL, HIDE_NOTIFICATION, SHOW_MODAL, SHOW_NOTIFICATION } from './constants'

export const showModal = createAction<ShowModalProps>(SHOW_MODAL)

export const hideModal = createAction(HIDE_MODAL)

export const showNotification = createAction(
  SHOW_NOTIFICATION,
  ({
    duration = NOTIFICATION_DURATION,
    messageText,
    messageType = NOTIFICATION_TYPE.SUCCESS
  }: ShowNotificationProps): { payload: INotification } => {
    const id = nanoid()
    return { payload: { duration, id, messageText, messageType } }
  }
)

export const hideNotification = createAction<string>(HIDE_NOTIFICATION)
