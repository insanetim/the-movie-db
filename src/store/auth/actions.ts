import Cookies from 'js-cookie'
import {
  createRequestToken,
  createSession,
  deleteSession,
  getAccountDetails,
  validateWithLogin,
} from 'src/api/tmdb/apiRoutes'
import { NOTIFICATION_TYPE } from 'src/constants/app'
import { showNotification } from 'src/store/app/actions'
import errorMessage from 'src/utils/helpers/errorMessage'
import getSessionId from 'src/utils/helpers/getSessionId'

import { createAppAsyncThunk } from '../withTypes'
import * as types from './constants'
import { UserData } from './types'

const logIn = createAppAsyncThunk(
  types.logIn,
  async function (userData: UserData, { dispatch }) {
    try {
      const requestToken = await createRequestToken()
      await validateWithLogin({ requestToken, userData })
      const sessionId = await createSession({ requestToken })

      Cookies.set('tmdb.session_id', sessionId)

      return sessionId
    } catch (error) {
      dispatch(
        showNotification({
          message: errorMessage(error),
          type: NOTIFICATION_TYPE.ERROR,
        })
      )
    }
  }
)

const logOut = createAppAsyncThunk(
  types.logOut,
  async function (_, { dispatch }) {
    const sessionId = getSessionId()

    try {
      await deleteSession({ sessionId })

      Cookies.remove('tmdb.session_id')
    } catch (error) {
      dispatch(
        showNotification({
          message: errorMessage(error),
          type: NOTIFICATION_TYPE.ERROR,
        })
      )
    }
  }
)

const fetchAccount = createAppAsyncThunk(
  types.fetchAccount,
  async function (_, { dispatch }) {
    const sessionId = getSessionId()

    try {
      const account = await getAccountDetails({ sessionId })

      return account
    } catch (error) {
      dispatch(
        showNotification({
          message: errorMessage(error),
          type: NOTIFICATION_TYPE.ERROR,
        })
      )
    }
  }
)

export { fetchAccount, logIn, logOut }
