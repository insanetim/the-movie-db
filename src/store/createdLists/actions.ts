import { createAsyncThunk } from '@reduxjs/toolkit'
import { NOTIFICATION_TYPE } from 'src/constants/app'
import { IList, IListsList } from 'src/interfaces/list.interface'
import {
  addMovieToList,
  createNewList,
  deleteMyList,
  getCreatedLists,
  removeMovieFromList,
} from 'src/services/api/apiRoutes'
import { showNotification } from 'src/store/app/actions'
import { accountSelector } from 'src/store/auth/selectors'
import errorMessage from 'src/utils/helpers/errorMessage'
import getSessionId from 'src/utils/helpers/getSessionId'
import listMessage from 'src/utils/helpers/listMessage'

import { RootState } from '../index'
import { selectMovieById } from '../movie/selectors'
import * as types from './constants'
import { AddToListProps, CreateListProps, RemoveFromListProps } from './types'

const fetchLists = createAsyncThunk<
  IListsList,
  string,
  { rejectValue: string; state: RootState }
>(types.fetchLists, async function (page, { getState, rejectWithValue }) {
  const sessionId = getSessionId()
  const accountId = accountSelector(getState())!.id

  try {
    const lists = await getCreatedLists({ accountId, page, sessionId })

    return lists
  } catch (error) {
    return rejectWithValue(errorMessage(error))
  }
})

const createList = createAsyncThunk<
  void,
  CreateListProps,
  { state: RootState }
>(types.createList, async function ({ listData, movieId }, { dispatch }) {
  const sessionId = getSessionId()
  const listName = listData.name

  try {
    const listId = await createNewList({ listData, sessionId })

    if (typeof movieId !== 'undefined') {
      dispatch(addToList({ listId, listName, movieId }))
    }
  } catch (error) {
    dispatch(
      showNotification({
        messageText: errorMessage(error),
        messageType: NOTIFICATION_TYPE.ERROR,
      })
    )
  }
})

const deleteList = createAsyncThunk<void, IList['id'], { state: RootState }>(
  types.deleteList,
  async function (listId, { dispatch }) {
    const sessionId = getSessionId()

    try {
      await deleteMyList({ listId, sessionId })
    } catch (error) {
      dispatch(
        showNotification({
          messageText: errorMessage(error),
          messageType: NOTIFICATION_TYPE.ERROR,
        })
      )
    }
  }
)

const addToList = createAsyncThunk<void, AddToListProps, { state: RootState }>(
  types.addToList,
  async function ({ listId, listName, movieId }, { dispatch, getState }) {
    const sessionId = getSessionId()
    const movieTitle = selectMovieById(getState(), movieId)!.title

    try {
      await addMovieToList({ listId, movieId, sessionId })
      dispatch(
        showNotification({ messageText: listMessage(movieTitle, listName) })
      )
      await dispatch(fetchLists('1'))
    } catch (error) {
      dispatch(
        showNotification({
          messageText: errorMessage(error),
          messageType: NOTIFICATION_TYPE.ERROR,
        })
      )
    }
  }
)

const removeFromList = createAsyncThunk<
  void,
  RemoveFromListProps,
  { state: RootState }
>(types.removeFromList, async function ({ listId, movieId }, { dispatch }) {
  const sessionId = getSessionId()

  try {
    await removeMovieFromList({ listId, movieId, sessionId })
  } catch (error) {
    dispatch(
      showNotification({
        messageText: errorMessage(error),
        messageType: NOTIFICATION_TYPE.ERROR,
      })
    )
  }
})

export { addToList, createList, deleteList, fetchLists, removeFromList }
