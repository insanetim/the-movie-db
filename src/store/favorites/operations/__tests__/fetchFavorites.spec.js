import mockHttpClient from 'src/__mocks__/httpClientMock'
import * as types from '../../types'
import { fetchFavoritesRequest, fetchFavoritesSuccess, fetchFavoritesFailure } from '../../actions'
import fetchFavorites from '../fetchFavorites'

jest.mock('src/store/session/selectors', () => ({
  sessionIdSelector: jest.fn(() => 'session_id'),
  accountSelector: jest.fn(() => ({ id: 123 }))
}))

describe('fetchFavorites', () => {
  const dispatch = jest.fn()

  const action = {
    type: types.FETCH_FAVORITES,
    payload: 1
  }

  const url = '/account/123/favorite/movies'
  const body = {
    params: { session_id: 'session_id', page: 1 }
  }
  const response = {
    data: {
      page: 1,
      results: [],
      total_pages: 0,
      total_results: 0
    }
  }

  const beforeFunction = httpClient => () => {
    fetchFavorites.process(
      {
        httpClient,
        action,
        getState: jest.fn()
      },
      dispatch,
      jest.fn()
    )
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('has valid attributes', () => {
    expect(fetchFavorites).toMatchSnapshot()
  })

  describe('success', () => {
    const httpClient = mockHttpClient({ method: 'get', response })

    beforeEach(beforeFunction(httpClient))

    it('calls right endpoint', () => {
      expect(httpClient.get).toHaveBeenCalledTimes(1)
      expect(httpClient.get).toHaveBeenCalledWith(url, body)
    })

    it('dispatches actions', () => {
      expect(dispatch).toHaveBeenCalledTimes(2)
      expect(dispatch).toHaveBeenNthCalledWith(1, fetchFavoritesRequest(1))
      expect(dispatch).toHaveBeenNthCalledWith(2, fetchFavoritesSuccess(response.data))
    })
  })

  describe('failure', () => {
    const error = new Error('test/error')

    const httpClient = mockHttpClient({
      method: 'get',
      reject: true,
      response: error
    })

    beforeEach(beforeFunction(httpClient))

    it('dispatches actions', () => {
      expect(dispatch).toHaveBeenCalledTimes(2)
      expect(dispatch).toHaveBeenNthCalledWith(1, fetchFavoritesRequest(1))
      expect(dispatch).toHaveBeenNthCalledWith(2, fetchFavoritesFailure(error))
    })
  })
})
