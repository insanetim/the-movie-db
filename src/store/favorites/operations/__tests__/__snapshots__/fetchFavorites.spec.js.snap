// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchFavorites has valid attributes 1`] = `
{
  "cancelType": undefined,
  "debounce": 0,
  "latest": true,
  "name": undefined,
  "process": [Function],
  "processOptions": {
    "dispatchMultiple": true,
  },
  "throttle": 0,
  "transform": undefined,
  "type": "watchlist/FETCH_FAVORITES",
  "validate": undefined,
  "warnTimeout": 60000,
}
`;
